src/crystalla/ndarray.cr

```diff
@@ -0,0 +1,523 @@
+    end
+
+    # broadcasting for division
+    def /(number : Number) : Ndarray
+      other = broadcast(number)
+      zip_with(other) { |x, y| x / y }
+    end
+
+    def - : Ndarray
+      Ndarray.new(values.map(&.-), @shape)
+    end
+
+    def **(number : Number) : Ndarray
+      other = broadcast(number)
+      zip_with(other) { |x, y| x ** y }
+    end
+
+    def zip_with(other) : Ndarray
+      # expand rows & columns
+      if shape[1] == 1 && (other.shape[0] == 1 || other.shape[0] == 0)
+        other = Ndarray.tile(other.values, @shape[0].to_i32)
+        self_values = [] of Float64
+        (1...self.shape[0]).each do |i|
+          self_values += self.values
+        end
+        expand_self = Ndarray.new(self_values, {shape[0], other.shape[1]})
+        new_values = Array.new(expand_self.values.size, 0.0)
+        expand_self.each do |i, x|
+          new_values[i] = yield x, other.values[i]
+        end
+        return Ndarray.new(new_values, expand_self.shape)
+
+      # expand rows
+      elsif shape[0] != other.shape[0] && (other.shape[0] == 1 || other.shape[0] == 0)
+        other = Ndarray.tile(other.values, @shape[0])
+      elsif shape[0] != other.shape[0]
+        raise ArgumentError.new "number of rows mismatch in array operation"
+      end
+
+
+      # expand columns
+      if shape[1] != other.shape[1] && other.shape[1] == 1
+        other_values = [] of Float64
+        (0..other.shape[0]).each do |i|
+          other_values += other.values
+        end
+        other = Ndarray.new(other_values, {shape[0], shape[1]})
+      elsif shape[1] != other.shape[1]
+        raise ArgumentError.new "number of columns mismatch in array operation"
+      end
+
+      new_values = Array.new(@values.size, 0.0)
+      self.each do |i, x|
+        new_values[i] = yield x, other.values[i]
+      end
+      Ndarray.new(new_values, shape)
+    end
+
+    # sum elements of the whole array, or following an axis
+    def sum(axis = nil)
+      if axis.nil?
+        self.values.reduce{ |acc, i| acc + i }
+      elsif axis == 0
+        arr = self.values.in_groups_of(shape[1])
+        sum = arr.map{|v| v.compact.sum(0.0)}.flatten
+        Ndarray.new(sum)
+      elsif axis == 1
+        sum = [] of Float64
+        self.each_row{|r| sum << r.sum}
+        Ndarray.new(sum)
+      else
+        raise ArgumentError.new "Axis provided must be 0(columns) or 1(rows)"
+      end
+    end
+
+    # "dot product": multiplication or arrays
+    def dot (other : self) : Ndarray
+      if other.shape[0] == 0 && shape[1] != other.shape[1]
+        raise ArgumentError.new "number of rows/columns mismatch in array multiplication"
+      elsif other.shape[0] != 0 && shape[1] != other.shape[0]
+        raise ArgumentError.new "number of rows/columns mismatch in array multiplication"
+      end
+
+      blas_multiply_array other
+    end
+
+    # reshape array C style or Fortran style
+    def reshape(number_of_rows, number_of_cols, order="F")
```

Comment:
I think it's better if we default to `order="C"`. Numpy does it like that, and most of the modern developers will be a bit alienated if they call `reshape(x, y)` and it's backwards.